x-pack/plugins/cross_cluster_replication/public/app/sections/home/auto_follow_pattern_list/auto_follow_pattern_list.js

```diff
@@ -21,17 +21,37 @@ export const AutoFollowPatternList = injectI18n(
   class extends PureComponent {
     static propTypes = {
       loadAutoFollowPatterns: PropTypes.func,
+      selectAutoFollowPattern: PropTypes.func,
       autoFollowPatterns: PropTypes.array,
       apiStatus: PropTypes.string,
       apiError: PropTypes.object,
-      openDetailPanel: PropTypes.func.isRequired,
-      closeDetailPanel: PropTypes.func.isRequired,
-      isDetailPanelOpen: PropTypes.bool,
     }
 
+    static getDerivedStateFromProps({ autoFollowPatternId }, { lastAutoFollowPatternId }) {
```

Comment:
Just a personal DX anecdote... as I read this code, I found myself having to trace the logic through three lifecycle methods: `componentDidMount`, `getDerivedStateFromProps`, and `componentDidUpdate`. I then had to also factor in what would happen when the user clicks a row (a new pattern is selected, triggering `getDerivedStateFromProps`). This took me a few minutes to wrap my head around, and then I was still left wondering why we're storing `lastAutoFollowPatternId`. Through experimentation by removing the logic it supports, I found that its purpose is to prevent an infinite render loop.

To be honest, I feel like this code has become more complex with this change. What is the benefit this complexity brings us?

Reply:
`ComponentDidMount` is pretty clear I think, it mounts and checks for the presence of a pattern in the query params. I understand what u say about the 2 other lifecycle methods, I will see how this could be improved.
There is never a benefit when complexity comes in 😊 I thought that using a middleware to change a URL on a single page was also a bit complex, mainly because you had to remember that it existed and that it was there that the search params were being changed. So I tried to move the logic as close as possible to the page (section) component.